tools/tsp-client/src/fs.ts

```
@@ -0,0 +1,89 @@
+  for (const [relativeFilePath, contents] of files) {
+    const filePath = path.join(rootDir, relativeFilePath);
+    await ensureDirectory(path.dirname(filePath));
+    Logger.debug(`writing ${filePath}`);
+    await writeFile(filePath, contents);
+  }
+}
+
+export async function tryReadTspLocation(rootDir: string): Promise<string | undefined> {
+  try {
+    const yamlPath = path.resolve(rootDir, "tsp-location.yaml");
+    const fileStat = await stat(yamlPath);
+    if (fileStat.isFile()) {
+      const fileContents = await readFile(yamlPath, "utf8");
+      const locationYaml = parseYaml(fileContents);
+      const { directory, commit, repo } = locationYaml;
+      if (!directory || !commit || !repo) {
+        throw new Error("Invalid tsp-location.yaml");
+      }
+      // make GitHub URL
+      return `https://raw.githubusercontent.com/${repo}/${commit}/${directory}/`;
+    }
+  } catch (e) {
+    Logger.error(`Error reading tsp-location.yaml: ${e}`);
+  }
+  return undefined;
+}
+
+export async function readTspLocation(rootDir: string): Promise<[string, string, string, string[]]> {
+  try {
+    const yamlPath = path.resolve(rootDir, "tsp-location.yaml");
+    const fileStat = await stat(yamlPath);
+    if (fileStat.isFile()) {
+      const fileContents = await readFile(yamlPath, "utf8");
+      const locationYaml = parseYaml(fileContents);
+      let { directory, commit, repo, additionalDirectories } = locationYaml;
+      if (!directory || !commit || !repo) {
+        throw new Error("Invalid tsp-location.yaml");
+      }
+      Logger.info(`Additional directories: ${additionalDirectories}`)
+      if (!additionalDirectories) {
+        additionalDirectories = [];
+      }
+      return [ directory, commit, repo, additionalDirectories ];
+    }
+  } catch (e) {
+    Logger.error(`Error reading tsp-location.yaml: ${e}`);
+  }
+  throw new Error("Could not find tsp-location.yaml");
```

maybe just rethrow the above exception in the catch instead?

---

I'm throwing here because the typing system didnt like that there was no return at the bottom of this function and I didnt want to add a `return undefined` here which would just make us check for undefined after the function returns then throw there. Let me know if there's something else that should be done here!

---

I think if you change the above to 

```ts
  } catch (e) {
    Logger.error(`Error reading tsp-location.yaml: ${e}`);
    throw e;
  }
```

or if you want to wrap the original error

```ts
  } catch (e) {
    Logger.error(`Error reading tsp-location.yaml: ${e}`);
    throw new Error(`Error reading tsp-location.yaml: ${e}`);
  }
```

Then you can move the `throw new Error("Could not find tsp-location.yaml");` line after the `if (fileStat.isFile()) {` check so it looks a little tidier. Really just a style nit, it felt odd to catch an error and log then immediately throw afterwards